src/platform/macos.rs

```diff
@@ -205,12 +206,34 @@ impl MonitorHandleExtMacOS for MonitorHandle {
     }
 }
 
+#[derive(Debug, PartialEq, Eq, Copy, Clone)]
+pub enum FileOpenResult {
+    Success,
+    Cancel,
+    Failure,
+}
+
 /// Additional methods on `EventLoopWindowTarget` that are specific to macOS.
 pub trait EventLoopWindowTargetExtMacOS {
     /// Hide the entire application. In most applications this is typically triggered with Command-H.
     fn hide_application(&self);
     /// Hide the other applications. In most applications this is typically triggered with Command+Option-H.
     fn hide_other_applications(&self);
+
+    /// Set or remove the callback which gets called when a file or files are requested to
+    /// be opened.
+    ///
+    /// This for example happens when the user double-clicks on a file in Finder and the
+    /// file's type is associated with this application.
+    ///
+    /// The callback function receives the absolute path to the target files as an argument and it should
+    /// indicate the success of the operation in its return value. See: `application:openFiles:` in AppKit.
+    ///
+    /// Other systems usually provide the path as a program argument.
+    /// See: `std::env::args`.
+    fn set_file_open_callback<F>(&self, callback: Option<F>)
```

Comment:
Great catch!

I'm seconding the creation of `EventLoopExtMacOS`, though it seems like it'd be useful to have the method on both?

Reply:
> I'm fairly certain this should be set on EventLoopExtMacOS

I like the intention to communicate that this should be called before calling `run`, but I agree with @francesca64 in that this should also be callable while the event loop is running. So I believe it should be implemented on `EventLoopWindowTarget`.

However I don't see why it would be beneficial to have it on both since `EventLoop` implements `Deref<Target=EventLoopWindowTarget>` and as such, all functions implemented on `EventLoopWindowTarget` can be invoked on `EventLoop` as well.

Reply:
I don't really see when you would call `set_file_open_callback`, since . Can you give an example of where this would be useful?

If not, I would probably argue it should be `new_with_file_open_callback` or similar which you call on creation of the EventLoop (or maybe have an `EventLoopBuilder` at some point in the future), such that you can't change the callback afterwards. This would probably also allow getting rid of a `Mutex` in the `Handler` later on.

Reply:
It would also mean you wouldn't need to create and store two delegate instances

Reply:
I have to admit that your arguments are really compelling. I'll wait a bit so that @francesca64 has time to respond, but otherwise I'm inclined to create that `new_with_file_open_callback` and remove `set_file_open_callback`.

Reply:
Seconding only having `new_with_file_open_callback`; it sounds like a concise and correct API!

Reply:
I hit a problem. The window needs an `EventLoop` when it's constructed. So `new_with_file_open_callback` must be called before the window exists, which leads to a very awkward solution if one desires to use anything depending on the event loop in the callback. The code below is the nicest approach I could come up with, and it's a mess. The `Mutex<Option<>>` part of the type wouldn't be needed if this could be called after the window was created, so that the closure could directly take a clone of `Arc<Window>`.
```rust
let window_ref: Arc<Mutex<Option<Window>>> = Default::default();
let event_loop;
{
    let window_ref = window_ref.clone();
    event_loop = EventLoop::<()>::new_with_file_open_callback(move |paths: Vec<PathBuf>| {
        let guard = window_ref.lock().unwrap();
        let window = match &*guard {
            Some(window) => window,
            None => return FileOpenResult::Failure,
        };
        for path in paths.iter() {
            open_file(window, path.as_ref());
        }
        FileOpenResult::Success
    });
}
``` 

Probably the best is to have a `set_file_open_callback` on the `EventLoop`.

Reply:
Hmm, I think I would argue that this behaviour is unidiomatic MacOS behaviour; one should almost always create new windows from the given files.

In the odd cases where you want to modify previously created windows, it is, as you've demonstrated, still possible, albeit requires a `Mutex` (which makes sense, since we're throwing a `Mutex` out in our `Handler`).

Though yeah, it dawns to me that we can't even access the event loop in `new_with_file_open_callback`...

Reply:
Also really dislike that we have to create two different delegates just for this functionality

Reply:
But I think I can accept it being like this for now (maybe with a few comments about what we've talked about here), at a later point `winit` really needs better callback support!

Reply:
> I think I can accept it being like this for now

Just to be clear, are you saying you'd be okay with having `set_file_open_callback` on the `EventLoop`?

Reply:
Yup, sorry for being unclear